File: Tyche.Android/data/account/src/main/java/com/felipearpa/data/account/type/Username.kt
package com.felipearpa.data.account.type

// * The username may only contain letters, numbers, underscores, and hyphens ([a-zA-Z0-9_-]).
// * The username must be between 3 and 16 characters in length ({3,16}).
private const val usernameRegularExpression = "^[a-zA-Z0-9_-]{3,16}\$"

@JvmInline
value class Username(val value: String) {
    init {
        checkEmpty(value)
        checkUsernamePattern(value)
    }

    override fun toString(): String = value

    companion object {

        fun isValid(value: String): Boolean {
            return try {
                checkEmpty(value)
                checkUsernamePattern(value)
                true
            } catch (_: IllegalArgumentException) {
                false
            }
        }

        private fun checkEmpty(value: String) {
            require(value.isNotEmpty()) { "Value must be not empty" }
        }

        private fun checkUsernamePattern(value: String) {
            val pattern = Regex(usernameRegularExpression)
            require(pattern.containsMatchIn(value)) { "Value must be a valid username" }
        }
    }
}

File: Tyche.Backend/Felipearpa.Account/Felipearpa.Account/Type/Username.fs
namespace Felipearpa.Account.Type

open System
open Felipearpa.Type.Validation

[<AutoOpen>]
module private UsernameHelper =

    // * The username may only contain letters, numbers, underscores, and hyphens ([a-zA-Z0-9_-]).
    // * The username must be between 3 and 16 characters in length ({3,16}).
    [<Literal>]
    let usernameRegularExpression =
        @"^[a-zA-Z0-9_-]{3,16}$"

type Username =
    private
    | Username of string

    member this.Value =
        match this with
        | Username value -> value

    override this.ToString() =
        match this with
        | Username value -> value

[<RequireQualifiedAccess>]
module Username =

    let private create element =
        match element with
        | Null -> Error "String must be not null"
        | NotMatches usernameRegularExpression -> Error "String must be a valid username"
        | _ -> Ok <| Username element

    let tryOf element =
        match create element with
        | Ok value -> Some value
        | Error _ -> None

    let newOf element =
        match create element with
        | Ok value -> value
        | Error message -> raise <| ArgumentException(message)

    let value (element: Username) = element.Value

    let toString (element: Username) = element.ToString()
